File: common/amundsen_common/models/feature.py
# Copyright Contributors to the Amundsen project.
# SPDX-License-Identifier: Apache-2.0

from typing import List, Optional, Dict, Any

import attr

from amundsen_common.models.user import User
from amundsen_common.models.badge import Badge
from amundsen_common.models.tag import Tag
from amundsen_common.models.table import ProgrammaticDescription
from marshmallow3_annotations.ext.attrs import AttrsSchema


@attr.s(auto_attribs=True, kw_only=True)
class ColumnItem:
    column_name: str
    column_type: str


class ColumnItemSchema(AttrsSchema):
    class Meta:
        target = ColumnItem
        register_as_scheme = True


@attr.s(auto_attribs=True, kw_only=True)
class DataSample:
    # Modeled after preview data model in FE
    columns: List[ColumnItem]
    data: List[Dict[str, Any]]
    error_text: str


class DataSampleSchema(AttrsSchema):
    class Meta:
        target = DataSample
        register_as_scheme = True


@attr.s(auto_attribs=True, kw_only=True)
class FeatureWatermark:
    key: Optional[str]
    watermark_type: Optional[str]
    time: str


class FeatureWatermarkSchema(AttrsSchema):
    class Meta:
        target = FeatureWatermark
        register_as_scheme = True


@attr.s(auto_attribs=True, kw_only=True)
class Feature:
    key: Optional[str] = attr.ib(default=None)
    name: str
    version: str  # ex: "1.2.0"
    status: Optional[str]
    feature_group: str
    entity: Optional[str]
    data_type: Optional[str]
    availability: List[str]
    description: Optional[str] = attr.ib(default=None)
    owners: List[User]
    badges: List[Badge]
    tags: List[Tag]
    programmatic_descriptions: List[ProgrammaticDescription]
    watermarks: List[FeatureWatermark]
    last_updated_timestamp: Optional[int]
    created_timestamp: Optional[int]


class FeatureSchema(AttrsSchema):
    class Meta:
        target = Feature
        register_as_scheme = True


@attr.s(auto_attribs=True, kw_only=True)
class FeatureSummary:
    key: str  # ex: test_feature_group_name/test_feature_name/1.2.0
    name: str
    version: str
    availability: List[str]
    entity: List[str]
    description: Optional[str] = attr.ib(default=None)
    badges: List[Badge]
    last_updated_timestamp: Optional[int]


class FeatureSummarySchema(AttrsSchema):
    class Meta:
        target = FeatureSummary
        register_as_scheme = True

File: frontend/amundsen_application/static/js/interfaces/Feature.ts
// Copyright Contributors to the Amundsen project.
// SPDX-License-Identifier: Apache-2.0

import { User } from './User';
import { Badge } from './Badges';
import { Tag } from './Tags';
import { ProgrammaticDescription } from './TableMetadata';

export interface FeatureMetadata {
  key: string;
  name: string;
  version: string;
  status: string;
  feature_group: string;
  entity?: string;
  data_type?: string;
  availability: string[];
  description: string;
  owners: User[];
  badges: Badge[];
  owner_tags?: Tag[];
  tags: Tag[];
  programmatic_descriptions: ProgrammaticDescription[];
  watermarks: FeatureWatermark[];
  stats: FeatureStats[];
  last_updated_timestamp: number;
  created_timestamp: number;
  partition_column?: string;
}

// TODO - duplicated with FeatureResource in Resources.ts. Might delete this.
export interface FeatureSummary {
  key: string;
  name: string;
  version: string;
  availability: string[];
  entity?: string[];
  description: string;
}

// TODO - Figure out if we can reuse ColumnStats
export interface FeatureStats {
  name: string;
}

export interface FeatureWatermark {
  key: string;
  watermark_type: string;
  time: string;
}

export interface FeaturePreviewQueryParams {
  feature_name: string;
  feature_group: string;
  version: string;
}

export interface FeatureCode {
  key: string;
  source: string;
  text: string;
}
